runtime/caml/roots.h

```diff
@@ -22,8 +22,8 @@
 #include "memory.h"
 
 typedef void (*scanning_action) (void*, value, value *);
-CAMLextern void (*caml_scan_roots_hook)(scanning_action, void*,
-                                        caml_domain_state*);
+typedef void (*scan_roots_hook) (scanning_action, void*, caml_domain_state*);
+CAMLextern _Atomic scan_roots_hook caml_scan_roots_hook;
```

Comment:
`camlatomic.h` does not assume that `stdatomic.h` to be present. See https://github.com/ocaml/ocaml/blob/trunk/runtime/caml/camlatomic.h. Is this breaking this assumption?

Reply:
Indeed, something has to be done.
The main issue is gcc 4.8 support which has no equivalent to the `_Atomic` keyword, and instead has atomic operations taking arbitrary pointer types.
I see two ways to fix this. The simplest one is to introduce a `CamlAtomic` which expands to `_Atomic` or nothing, with the following two changes:
- Under gcc 4.8 define atomics as aliases for their underlying type, instead of structs. The current definition as a struct might catch more type errors, but it might be less important to support the developer experience under gcc 4.8, than supporting legacy systems running gcc 4.8 themselves.
- Don't treat C++ specially, you can include `<stdatomic.h>` from C++. Is there a reason for redefining `stdatomic.h` primitives from the C++ `atomic` header currently?

There is a second solution which is a bit more complicated so I would try to avoid it at first.

Reply:
Is there a CI that currently tests gcc 4.8? The `_Atomic` keyword is used liberally in `lfskiplist.h`; according to issues on other github projects this is not supposed to work with gcc 4.8.

Reply:
I think we're moving towards not supporting GCC 4.8 and requiring proper C11 support, see #10940.  MSVC support will be another can of worms... to be opened much later.

For the time being, let's use `_Atomic` types freely, we have more urgent concerns.

Reply:
> Is there a CI that currently tests gcc 4.8? 

One of the Jenkins CI machines runs CentOS 7, which has the problematic GCC 4.8 if I remember correctly, but this part of the CI is currently turned off waiting for #10913 to be resolved.

Reply:
I’m ok with requiring proper C11 support.